File: cache/src/strategies/lru/lruStrategy.js
const EvictionStrategy = require("../evictionStrategy");
const DoublyLinkedList = require("./doublyLinkedList");
const Node = require("./node");

class LRUStrategy extends EvictionStrategy {
  constructor(capacity) {
    super();
    this.capacity = capacity;
    this.map = new Map();
    this.list = new DoublyLinkedList();
  }
  get(key) {
    if (!this.map.has(key)) return -1;
    const node = this.map.get(key);
    this._moveToFront(node);
    return node.value;
  }

  put(key, value) {
    if (this.map.has(key)) {
      const node = this.map.get(key);
      node.value = value;
      this._moveToFront(node);
      return;
    }

    this._evict(); // if needed

    const node = new Node(key, value);
    this.map.set(key, node);
    this.list.addToFront(node);
  }

  _evict() {
      if(this.map.size >= this.capacity){
        const lru = this.list.removeLRU();
        this.map.delete(lru.key);
      }
  }

  _moveToFront(node) {
    this.list.removeNode(node);
    this.list.addToFront(node);
  }
}

module.exports = LRUStrategy;

File: javacache/src/main/java/org/example/strategy/lru/LRUStrategy.java
package org.example.strategy.lru;



import org.example.strategy.EvictionStrategy;
import org.example.strategy.StrategyOptions;

import java.util.HashMap;

public class LRUStrategy implements EvictionStrategy {

    private int capacity;
    private HashMap<Integer, Node> map;
    private DoublyLinkedList list;

    public LRUStrategy(int capacity) {
        this.capacity = capacity;
        this.map = new HashMap<>();
        this.list = new DoublyLinkedList();
    }

    @Override
    public int get(int key) {
        if (!map.containsKey(key)) {
            return -1;
        }
        Node n = map.get(key);
        list.addToFront(n);
        return n.value;
    }

    @Override
    public void put(int key, int value, StrategyOptions options) {
        if (map.containsKey(key)) {
            Node n = map.get(key);
            n.value = value;
            map.put(key, n);
            list.addToFront(n);
            return;
        }

        // eviction
        if (capacity <= map.size()) {
            Node lru = list.removeLRU();
            if (lru != null) {
                map.remove(lru.key);
            }
        }

        Node n = new Node(key, value);
        map.put(key, n);
        list.addToFront(n);
    }
}
